activerecord/lib/active_record/reflection.rb

```diff
@@ -175,8 +176,19 @@ def join_keys(association_klass)
         JoinKeys.new(foreign_key, active_record_primary_key)
       end
 
+      # Returns a list of scopes that should be applied for this Reflection
+      # object when querying the database.
+      def scopes
+        scope ? [scope] : []
+      end
+
+      def scope_chain
+        chain.map(&:scopes)
+      end
+      deprecate :scope_chain
+
       def constraints
-        scope_chain.flatten
+        chain.map(&:scopes).flatten
```

Comment:
Is better to use `flat_map` in stead of `map().flatten`.

http://apidock.com/ruby/Enumerable/flat_map